rpc/RPCSession.java

```diff
@@ -65,66 +74,282 @@ public RPCSession(GraknClient client, String database, Type type, GraknOptions o
+        }
+
+        private DatabaseReplicas discoverDatabaseReplicas() {
+            int attempt = 0;
+            while (attempt < clusterClient.coreClients().size()) {
+                try {
+                    return DatabaseReplicas.ofProto(
+                            clusterDiscoveryRPC.databaseReplicasDiscovery(
+                                    DiscoveryProto.Discovery.DatabaseReplicas.Req.newBuilder()
+                                            .setDatabase(database)
+                                            .build()
+                            )
+                    );
+                } catch (StatusRuntimeException e) {
+                    ClusterGrpc.ClusterBlockingStub old = clusterDiscoveryRPC;
+                    clusterDiscoveryRPC = this.clusterClient.selectNextClusterDiscoveryRPC();
+                    LOG.info("Cluster Discovery RPC failed - server '{}' not available. Attempting to perform the operation onto another server '{}'",
+                            old.getChannel().authority(), clusterDiscoveryRPC.getChannel().authority());
+                }
+            }
+
+            throw new GraknClientException(CLUSTER_NOT_AVAILABLE.message(clusterClient.coreClients()));
+        }
+
+        public static class DatabaseReplicas {
+            private final ConcurrentMap<DatabaseReplica.Id, DatabaseReplica> replicaMap;
+            private final DatabaseReplica[] replicaArray;
+            private final AtomicInteger selectedReplica;
+
+            private DatabaseReplicas(ConcurrentMap<DatabaseReplica.Id, DatabaseReplica> replicaMap) {
+                this.replicaMap = replicaMap;
+                replicaArray = replicaMap.values().toArray(new DatabaseReplica[]{});
+                selectedReplica = new AtomicInteger(0);
+            }
+
+            public static DatabaseReplicas ofProto(DiscoveryProto.Discovery.DatabaseReplicas.Res res) {
+                ConcurrentMap<DatabaseReplica.Id, DatabaseReplica> replicaMap = new ConcurrentHashMap<>();
+
+                for (DiscoveryProto.Discovery.DatabaseReplicas.Res.DatabaseReplica replica: res.getReplicasList()) {
+                    DatabaseReplica.Id id = new DatabaseReplica.Id(Address.Cluster.Server.parse(replica.getAddress()), replica.getDatabase());
+                    replicaMap.put(id, DatabaseReplica.ofProto(replica));
+                }
+
+                return new DatabaseReplicas(replicaMap);
+            }
+
+            private DatabaseReplica selectLeader() {
+                Map.Entry<DatabaseReplica.Id, DatabaseReplica> initial = replicaMap.entrySet().iterator().next();
+                Map.Entry<DatabaseReplica.Id, DatabaseReplica> reduce = replicaMap.entrySet().stream()
+                        .filter(entry -> entry.getValue().role())
+                        .reduce(initial, (acc, e) -> e.getValue().term > acc.getValue().term ? e : acc);
+                if (reduce.getValue().role()) return reduce.getValue();
```

Comment:
I think this line might be slightly magical. Does `role()` mean `is leader`? Could we refactor it to a property named `isLeader()`?

Reply:
Fixed in 7386886d9feb76543e9395176cc436f70726f00b